src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -0,0 +1,772 @@
+    new_snapshot->set(f_timestamp_ms, ms.count());
+
+    auto parent_snapshot = getParentSnapshot(parent_snapshot_id);
+    Poco::JSON::Object::Ptr summary = new Poco::JSON::Object;
+    summary->set("operation", "append");
+    summary->set(f_added_data_files, added_files);
+    summary->set(f_added_records, added_records);
+    summary->set(f_added_files_size, added_files_size);
+    summary->set(f_changed_partition_count, num_partitions);
+
+    auto sum_with_parent_snapshot = [&](const char * field_name, Int32 snapshot_value)
+    {
+        Int32 prev_value = parent_snapshot ? parent_snapshot->getObject(f_summary)->getValue<Int32>(field_name) : 0;
+        summary->set(field_name, prev_value + snapshot_value);
+    };
+
+    sum_with_parent_snapshot(f_total_records, added_records);
+    sum_with_parent_snapshot(f_total_files_size, added_files_size);
+    sum_with_parent_snapshot(f_total_data_files, added_files);
+    sum_with_parent_snapshot(f_total_delete_files, 0);
+    sum_with_parent_snapshot(f_total_position_deletes, 0);
+    sum_with_parent_snapshot(f_total_equality_deletes, 0);
+    new_snapshot->set(f_summary, summary);
+
+    new_snapshot->set(f_schema_id, parent_snapshot ? parent_snapshot->getValue<Int32>(f_schema_id) : 0);
+    new_snapshot->set(f_manifest_list, manifest_list_name);
+
+    metadata_object->getArray(f_snapshots)->add(new_snapshot);
+    metadata_object->set(Iceberg::f_current_snapshot_id, snapshot_id);
+    return new_snapshot;
+}
+
+ChunkPartitioner::ChunkPartitioner(
+    Poco::JSON::Array::Ptr partition_specification, Poco::JSON::Object::Ptr schema, ContextPtr context, const Block & sample_block_)
+    : sample_block(sample_block_)
+{
+    std::unordered_map<Int32, String> id_to_column;
+    {
+        auto schema_fields = schema->getArray(Iceberg::f_fields);
+        for (size_t i = 0; i < schema_fields->size(); ++i)
+        {
+            auto field = schema_fields->getObject(static_cast<UInt32>(i));
+            id_to_column[field->getValue<Int32>(Iceberg::f_id)] = field->getValue<String>(Iceberg::f_name);
+        }
+    }
+    for (size_t i = 0; i != partition_specification->size(); ++i)
+    {
+        auto partition_specification_field = partition_specification->getObject(static_cast<UInt32>(i));
+
+        auto transform_name = partition_specification_field->getValue<String>("transform");
+        transform_name = Poco::toLower(transform_name);
+
+        FunctionOverloadResolverPtr transform;
+
+        auto source_id = partition_specification_field->getValue<Int32>(Iceberg::f_source_id);
+        auto column_name = id_to_column[source_id];
+
+        std::optional<size_t> transform_param;
+        auto & factory = FunctionFactory::instance();
+
+        if (transform_name == "year" || transform_name == "years")
+            transform = factory.get("toYearNumSinceEpoch", context);
+
+        if (transform_name == "month" || transform_name == "months")
```

Comment:
Maybe we can put this code close to https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/ObjectStorage/DataLakes/Iceberg/ManifestFilesPruning.cpp#L34.